src/WebJobs.Script.WebHost/Middleware/HstsConfigurationMiddleware.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.HttpsPolicy;
+using Microsoft.Azure.WebJobs.Script.WebHost.Configuration;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.WebJobs.Script.Middleware
+{
+    public class HstsConfigurationMiddleware : IJobHostHttpMiddleware
+    {
+        private readonly IOptions<HostHstsOptions> _hostHstsOptions;
+        private HstsMiddleware _hstsMiddleware;
+        private RequestDelegate _next;
+        private double _nextConfigured = 0;
+        private InvokeDelegate _invoke;
+
+        public HstsConfigurationMiddleware(IOptions<HostHstsOptions> hostHstsOptions)
+        {
+            _hostHstsOptions = hostHstsOptions;
+            _invoke = InvokeIntialization;
+        }
+
+        private delegate Task InvokeDelegate(HttpContext httpContext, RequestDelegate next);
+
+        public async Task Invoke(HttpContext context, RequestDelegate next)
+        {
+            await _invoke(context, next);
+        }
+
+        public async Task InvokeIntialized(HttpContext context, RequestDelegate next)
+        {
+            await _next(context);
+        }
+
+        private async Task InvokeIntialization(HttpContext httpContext, RequestDelegate next)
+        {
+            if (Interlocked.CompareExchange(ref _nextConfigured, 1, 0) == 0)
+            {
+                if (_hostHstsOptions.Value.IsEnabled)
+                {
+                    _hstsMiddleware = new HstsMiddleware(next, _hostHstsOptions);
+                    Interlocked.Exchange(ref _next, _hstsMiddleware.Invoke);
+                }
+                else
+                {
+                    Interlocked.Exchange(ref _next, next);
+                }
+
+                Interlocked.Exchange(ref _invoke, InvokeIntialized);
+            }
+            await _invoke(httpContext, _next);
```

Comment:
There's a race here -- if a bunch of requests come in and check `_nextConfigured`, one of them will win and enter the "initialization" section here. The rest will flow through to `_invoke(httpContext, _next)`, which may just loop right back through to `InvokeInitialization` over-and-over until `_invoke` is set up. Is that intended?

I haven't thought through it fully, but could you make `_invoke` a `Lazy<InvokeDelegate>`?

Reply:
Yeah you are right about the race condition. Adding `await _invoke(httpContext, _next)` in the else statement might work, not entirely sure though. However we may not need this entirely. I was talking to @fabiocav to verify my understanding of `Interlocked.CompareExchange`. He pointed out that the code itself has a flaw in logic. I am using the `RequestDelegate next` passed to the invoke the method on the first invocation. However the it could be different in every request. Which will effectively break this piece of middleware. Would have to wire this up in a different manner.